internal/app/mapper/mapper.go

```diff
@@ -192,9 +195,53 @@ func isNotMatch(matchPredicate *matchPredicate, typeURL string, node *core.Node)
 	return !isMatch, nil
 }
 
-func getResult(fragmentRule *rule) string {
-	stringFragment := fragmentRule.GetResult().GetStringFragment()
-	return stringFragment
+func getResult(fragmentRule *rule, node *core.Node) (string, error) {
+	found, result, err := getResultFromRequestNodeFragment(fragmentRule, node)
+	if err != nil {
+		return "", err
+	}
+	if found {
+		return result, nil
+	}
+
+	return fragmentRule.GetResult().GetStringFragment(), nil
+}
+
+func getResultFromRequestNodeFragment(fragmentRule *rule, node *core.Node) (bool, string, error) {
+	requestNodeFragment := fragmentRule.GetResult().GetRequestNodeFragment()
+	if requestNodeFragment == nil {
+		return false, "", nil
+	}
+
+	nodeField := requestNodeFragment.GetField()
+	var nodeValue = ""
+	switch nodeField {
+	case aggregationv1.NodeFieldType_NODE_CLUSTER:
+		nodeValue = node.GetCluster()
+	case aggregationv1.NodeFieldType_NODE_ID:
+		nodeValue = node.GetId()
+	case aggregationv1.NodeFieldType_NODE_LOCALITY_REGION:
+		nodeValue = node.GetLocality().GetRegion()
+	case aggregationv1.NodeFieldType_NODE_LOCALITY_ZONE:
+		nodeValue = node.GetLocality().GetZone()
+	case aggregationv1.NodeFieldType_NODE_LOCALITY_SUBZONE:
+		nodeValue = node.GetLocality().GetSubZone()
+	}
+
+	action := requestNodeFragment.GetAction()
+	if action.GetExact() {
+		return true, nodeValue, nil
+	}
+
+	regexAction := action.GetRegexAction()
+	pattern := regexAction.GetPattern()
+	replace := regexAction.GetReplace()
+
+	reg, err := regexp.Compile(pattern)
+	if err != nil {
+		return false, "", err
+	}
+	return true, reg.ReplaceAllString(nodeValue, replace), nil
```

Comment:
If the replace result is empty, I think we should error, `No empty fragments allowed`. I think this will also make it simpler to get rid of the return bool. Assuming you had it to differentiate between empty string case and no result case.

Reply:
does this mean that we should have a tighter validation in https://github.com/envoyproxy/xds-relay/blob/master/api/protos/aggregation/v1/aggregation.proto#L123

Reply:
i think replace string can still be empty and produce a non empty fragment.